Add fleet service to login to AWS ECR.

cdmurph32 opened the pull request:
Add fleet service to login to AWS ECR. 
**This change also starts docker-cleanup.timer.**  I'm not sure if that was
intentionally deactivated in the switch from v1 to v2.

Please let me know if it would be better for the AWS ECR service and
timer file to live in fleet-local/worker.

cdmurph32 committed:
Add fleet service to login to AWS ECR.

This change also starts docker-cleanup.timer.  I'm not sure if that was
intentionally deactivated in the switch from v1 to v2.

Please let me know if it would be better for the AWS ECR service and
timer file to live in fleet-local/worker.

cdmurph32 committed:
Fix mistake.  Forgot to replace .hud with .ecr.

cdmurph32 committed:
Read ECR account and region from etcd.

chr0n1x commented on init:
```diff
@@ -42,7 +42,7 @@ do
 done
 # general-purpose units
-for unit in $(ls ${SCRIPTDIR}/v2/fleet|grep -e '.service$')
+for unit in $(ls ${SCRIPTDIR}/v2/fleet|grep -e '.service$\|.timer$')
```
oh dang, good catch!

chr0n1x commented on v2/fleet/aws-ecr-login.service:
```diff
@@ -0,0 +1,22 @@
+ExecStartPre=-/usr/bin/docker kill ecr-login
+ExecStartPre=-/usr/bin/docker rm -f ecr-login
+ExecStart=/bin/bash -c \
+    'source /etc/profile.d/etcdctl.sh && if [[ -n "$($AWS_ACCOUNT_ID)\" ]]; then eval $(/usr/bin/docker run --name ecr-login behance/utility aws ecr get-login --registry-ids $($AWS_ACCOUNT_ID) --region $($REGION)); fi'
```
a slight nitpick, but if possible split this into more lines.

chr0n1x commented on v2/fleet/aws-ecr-login.timer:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Runs Amazon ECR Login every 6 hours
```
why is this needed? maybe add some comments

cdmurph32 replied:
The ECR login is only valid for 12 hours.  So I created a timer to repeat refresh the credentials.  The instances seem to reboot fairly frequently so it might not be strictly necessary.

chr0n1x commented on v2/setup/ecr.sh:
```diff
@@ -0,0 +1,15 @@
+# /KEY/NAMESPACE VALUE
+while read line || [[ -n "$line" ]]; do
+    etcdctl set $line
+done < ${HOMEDIR}/.ecr
```
please add the content format to the [README](https://github.com/adobe-platform/mesos-systemd#files-in-s3)

cdmurph32 committed:
Separate ECR login into more lines, add comments.

aws-ecr-login.timer is necessary because the docker login credentials
are only valid for 12 hours.

cdmurph32 committed:
Merge branch 'master' into feature/ecr-login

cdmurph32 committed:
Update README.md

Add documentation for ECR login.

iderdik commented on v2/fleet/aws-ecr-login.service:
```diff
@@ -0,0 +1,26 @@
+[Service]
+Type=oneshot
+User=core
+Environment="AWS_ACCOUNT_ID=etcdctl get /ECR/registry-account"
```
Can we use an ECR-specific name for this env var so it's clear what we're using it for? (i.e. ECR_AWS_ACCOUNT_ID)

cdmurph32 committed:
Change variable name in service file.

AWS_ACCOUNT_ID is now ECR_AWS_ACCOUNT_ID.

cdmurph32 committed:
Write credentials to /home/core/.dockercfg.

The only credentials file mounted within the mesos slave container is
/home/core/.dockercfg.  The previous method failed because newer docker
versions write credentials to ~/.docker/config.json instead of
~/.dockercfg.

This forced me to find a more robust approach to the creation of
.dockercfg since aws cli would not work as it only spits out the 'docker
login'.  Fortunately I found a solution in a github project and
dockerhub image, which I forked and added onto.

Details can be found here https://github.com/adobe-platform/ecr-login.
The tool fetches the credentials and places them in the .dockercfg
format.  The aws-ecr-login service overwrites /home/core/.dockercfg with
the output.

The .dockercfg for the behance dockerhub registry remain in
/root/.dockercfg and are used by fleetctl/systemd services even when
other users are specified in the service files.  Therefore this does not
interfere with other services that my need those credentials.  It will
interfere with any containers run through Marathon that are not found in
the chosen ECR registry.  If desired, this challenge could be
circumvented by allowing the specification of the Marathon URI field in
FlightDirector.  The URI could be directed at the URL of an S3 bucket
and secured with the S3 container policy.  This could be automated with
an additional fleetctl service which uses the ecr-login container
with the added modification of s3 push.

The alternative is to have all images exist in a single ECR registry.
IMHO the best way to manage this is for PDF Services build automation to
push to that ECR registry.  It's very simple to adjust the permissions on
a per repository basis in ECR.

That discussion aside, when we cease using DockerHub the ecr-login
Docker image will need to pulled using alternate means.  The simplest
way I can think of is to store it as a tarball in an S3 bucket which is
pulled by the worker instances at boot.
